enhancements/network/service-egress-traffic-steering.md

```diff
@@ -233,9 +243,10 @@ metadata:
   name: example-service
   namespace: some-namespace
 spec:
-  nodeSelector:
-    matchLabels:
-      color: green
+  snat:
+    nodeSelector:
```

Comment:
so what if as a user I dont want to snat, but only some of my nodes have NICs to a specific network i want to use? Shouldn't I still be able to use nodeSelector to only egress traffic out of those nodes?

Reply:
added a `Network without SNAT` section, here it is the user's responsibility to make sure the pods run on the correct nodes, because the controller will not allocate/designate other nodes to take care of pods that are not local to them.

Reply:
but it potentially could right? we are just not implementing it. What if someone asks for this feature in the future, then our CRD doesn't make much sense with it being a subfield of snat?

Reply:
I'm not sure if there's much value in selecting multiple nodes for the no-snat scenario as it feels more natural to require eps to be scheduled to nodes that have the network instead of trying to make everything work someway (covering all of the corner-cases).
doing this it doesn't make sense to separate snat specific fields instead of bundling them together, like opening users to set nodeselector without snat might suggest things we don't (or ever will) support. In the future we might add more snat specific fields and they would fit naturally inside the snat struct. (this was also discussed in the meeting I couldn't attend :sweat_smile: )

Reply:
I think it makes sense because we are talking here about traffic leaving a node on separate networks/interfaces, which only a subset of nodes may actually be connected to. So if I have a few of my nodes in a region that has connectivity to network x, i want to only select those to be egress for that network.

Reply:
in that case I think it would be much simpler to just schedule the pods only on these nodes (e.g by using the Deployment's nodeSelector) rather than doing some complicated egress service node selection + steering to make it work.

Reply:
OK if you want to limit the functionality to that, it's fine. But I still don't like boxing ourselves into a corner with this API. Why do you even need a field structure like this. Why not just make it:

```yaml
lb-snat-node-selector:
   matchLabels:
       color: green
```

Reply:
because the field structure enables the user to request snat without a selector in an easy way (snat:{}). having a field like you're suggesting means we need another field for enabling snat and then the additional selector field depends on that one being present. it feels cleaner to just bundle them together in one field (and if we add more criteria in the future they'll fit naturally in the sub-field) as the hierarchy makes sense imo.

Reply:
I see. I still don't really like the generic name "snat" and the structure. Adding @jcaamano in case he has some ideas/opinons.

Reply:
The more natural term for me here is `masquerade` instead of `snat`. It is still a technical term but I think it is good enough.

And I have the feeling this is the natural expression of the API:
```
network: ...
masquerade: true
nodeSelector: ...
```

The only reason not to do it this way is because we think there is a better way to achieve no-snat on specific node but that forces out to a less natural expression of the API.

Whereas we could either:

- Support that natural expression of the API
- Not support it via validation & status reporting instead of going for the less natural expression and forcing us out of supporting it in the future.

Reply:
I also think that going for implicitness in APIs (empty values having an special meaning) is not the best choice.

Reply:
I agree with @jcaamano. I think we should go with that structure and then in validation if someone tries to use nodeSelector with masquerade:false we considerate invalid either via kubeapi server validation (not sure if we can list invalid combinations of fields) or with ovnk posting a status invalid to the CRD or something.

Reply:
done (dropped the struct),  I feel `masquerade` suffers from a similar vagueness as `snat` so went with `snatToLoadBalancerIP` as I think it captures the meaning a bit better, wdyt?